File: backend/app/schemas/leisure.py
"""Pydantic schemas for Leisure Budget v2.0.

Three resources:
* ``LeisureActivity`` — user-owned pool entries.
* ``LeisureSession`` — completed leisure sessions (timer or manual).
* ``LeisureSettings`` — per-user daily target + enforcement config.

Categories + enforcement modes are constrained to the spec-locked
enum values; the request validators reject anything else with a 422.
"""

from datetime import date as date_cls, datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator


LeisureCategoryT = Literal["PHYSICAL", "SOCIAL", "CREATIVE", "PASSIVE"]
LeisureEnforcementModeT = Literal["SOFT", "MEDIUM", "HARD"]
LeisureSourceT = Literal["TIMER", "MANUAL"]


class LeisureActivityCreate(BaseModel):
    id: str = Field(..., min_length=1, max_length=64)
    name: str = Field(..., min_length=1, max_length=120)
    category: LeisureCategoryT
    default_duration_minutes: Optional[int] = Field(None, ge=1, le=1440)
    enabled: bool = True


class LeisureActivityUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    category: Optional[LeisureCategoryT] = None
    default_duration_minutes: Optional[int] = Field(None, ge=1, le=1440)
    enabled: Optional[bool] = None


class LeisureActivityResponse(BaseModel):
    id: str
    name: str
    category: LeisureCategoryT
    default_duration_minutes: Optional[int] = None
    enabled: bool
    created_at: datetime
    updated_at: datetime
    last_completed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class LeisureSessionCreate(BaseModel):
    id: str = Field(..., min_length=1, max_length=64)
    activity_id: Optional[str] = Field(None, max_length=64)
    category: LeisureCategoryT
    duration_minutes: int = Field(..., ge=1, le=1440)
    logged_at: datetime
    source: LeisureSourceT


class LeisureSessionResponse(BaseModel):
    id: str
    activity_id: Optional[str] = None
    category: LeisureCategoryT
    duration_minutes: int
    logged_at: datetime
    source: LeisureSourceT
    created_at: datetime

    model_config = {"from_attributes": True}


class LeisureSettingsResponse(BaseModel):
    """The effective leisure settings the client should display.

    ``pending_enforcement_mode`` may be set when the client just changed
    enforcement_mode but it hasn't taken effect yet — display as
    "Effective tomorrow" in the UI. The currently-active mode is the
    ``enforcement_mode`` field; clients should NOT treat ``pending`` as
    active.
    """

    daily_target_minutes: int
    weekend_target_minutes: Optional[int] = None
    enforcement_mode: LeisureEnforcementModeT
    refresh_limit: int
    enabled_categories: list[LeisureCategoryT]
    pending_enforcement_mode: Optional[LeisureEnforcementModeT] = None
    pending_enforcement_effective_date: Optional[date_cls] = None
    updated_at: datetime


class LeisureSettingsUpdate(BaseModel):
    daily_target_minutes: Optional[int] = Field(None, ge=0, le=1440)
    weekend_target_minutes: Optional[int] = Field(None, ge=0, le=1440)
    # Setting enforcement_mode to anything other than SOFT requires Pro
    # tier — gated server-side by ``require_leisure_enforcement_choice``.
    enforcement_mode: Optional[LeisureEnforcementModeT] = None
    refresh_limit: Optional[int] = Field(None, ge=0, le=10)
    enabled_categories: Optional[list[LeisureCategoryT]] = None
    # Set by the client on day-rollover when promoting a pending
    # enforcement-mode change to active. The server clears
    # pending_enforcement_mode + pending_enforcement_effective_date when
    # this is set to true.
    promote_pending_enforcement: Optional[bool] = None

    @field_validator("enabled_categories")
    @classmethod
    def _at_least_one_category(
        cls, v: Optional[list[str]]
    ) -> Optional[list[str]]:
        if v is not None and len(v) == 0:
            # Empty would make the random-pull algorithm return nothing
            # forever — block that at the API boundary rather than have
            # the UI ship into a dead state.
            raise ValueError("at least one category must be enabled")
        return v

File: web/src/types/leisure.ts
/**
 * Leisure Budget v2.0 types — mirrors the FastAPI Pydantic schemas at
 * `backend/app/schemas/leisure.py` and the Android Room entities
 * `LeisureActivityEntity` / `LeisureSessionEntity`.
 *
 * Categories are spec-locked server-side to the four built-in buckets via a
 * CHECK constraint; user-defined custom categories stay client-side only (the
 * Android counterpart lives in `LeisureBudgetPreferences` DataStore under the
 * `custom:<uuid8>` id namespace). Custom-tagged activities/sessions do NOT
 * round-trip through the REST API and are filtered out at the network
 * boundary by `leisureApi.createActivity` / `createSession`.
 */

export type LeisureCategory = 'PHYSICAL' | 'SOCIAL' | 'CREATIVE' | 'PASSIVE';
export type LeisureEnforcementMode = 'SOFT' | 'MEDIUM' | 'HARD';
export type LeisureSessionSource = 'TIMER' | 'MANUAL';

export const LEISURE_CATEGORIES: readonly LeisureCategory[] = [
  'PHYSICAL',
  'SOCIAL',
  'CREATIVE',
  'PASSIVE',
] as const;

/**
 * Display metadata for the four built-in categories. Custom-category labels
 * + emojis are stored alongside the id in `leisureStore.customCategories`
 * (later PR). Mirrors `LeisureCategory.kt`.
 */
export const LEISURE_CATEGORY_DISPLAY: Record<
  LeisureCategory,
  { emoji: string; label: string }
> = {
  PHYSICAL: { emoji: '🏃', label: 'Physical' },
  SOCIAL: { emoji: '👥', label: 'Social' },
  CREATIVE: { emoji: '🎨', label: 'Creative' },
  PASSIVE: { emoji: '🛋️', label: 'Passive' },
};

export interface LeisureActivity {
  id: string;
  name: string;
  category: LeisureCategory;
  default_duration_minutes: number | null;
  enabled: boolean;
  created_at: string;
  updated_at: string;
  last_completed_at: string | null;
}

export interface LeisureActivityCreate {
  id: string;
  name: string;
  category: LeisureCategory;
  default_duration_minutes?: number | null;
  enabled?: boolean;
}

export interface LeisureActivityUpdate {
  name?: string;
  category?: LeisureCategory;
  default_duration_minutes?: number | null;
  enabled?: boolean;
}

export interface LeisureSession {
  id: string;
  activity_id: string | null;
  category: LeisureCategory;
  duration_minutes: number;
  logged_at: string;
  source: LeisureSessionSource;
  created_at: string;
}

export interface LeisureSessionCreate {
  id: string;
  activity_id?: string | null;
  category: LeisureCategory;
  duration_minutes: number;
  logged_at: string;
  source: LeisureSessionSource;
}

export interface LeisureSettings {
  daily_target_minutes: number;
  weekend_target_minutes: number | null;
  enforcement_mode: LeisureEnforcementMode;
  refresh_limit: number;
  enabled_categories: LeisureCategory[];
  pending_enforcement_mode: LeisureEnforcementMode | null;
  pending_enforcement_effective_date: string | null;
  updated_at: string;
}

export interface LeisureSettingsUpdate {
  daily_target_minutes?: number;
  weekend_target_minutes?: number | null;
  enforcement_mode?: LeisureEnforcementMode;
  refresh_limit?: number;
  enabled_categories?: LeisureCategory[];
  promote_pending_enforcement?: boolean;
}

/**
 * Parse a stored category string back into a known LeisureCategory, or null
 * if the value is a custom-namespace id or otherwise unrecognized.
 */
export function parseLeisureCategory(
  value: string | null | undefined,
): LeisureCategory | null {
  if (!value) return null;
  return (LEISURE_CATEGORIES as readonly string[]).includes(value)
    ? (value as LeisureCategory)
    : null;
}

export const CUSTOM_CATEGORY_PREFIX = 'custom:';

export function isCustomCategoryId(value: string | null | undefined): boolean {
  return typeof value === 'string' && value.startsWith(CUSTOM_CATEGORY_PREFIX);
}
